src/main/scala/beam/utils/map/R5MapStatsCalculator.scala

```diff
@@ -1,24 +1,170 @@
+    // LoggerFactory.getLogger("com.conveyal").asInstanceOf[Logger].setLevel(Level.INFO)
+    val pathToOsm = "/mnt/data/work/beam/beam-production/production/sfbay/r5-updated/sf-bay.osm.pbf" // args(0)
 
+    analyzeR5Map(pathToOsm)
+    analyzeOSMMap(pathToOsm)
+  }
+
+  private def analyzeOSMMap(pathToOsm: String): Unit = {
+    val osm = new OSM(null)
+    try {
+      osm.readFromFile(pathToOsm)
+
+      println(s"Number of OSM nodes: ${osm.nodes.size()}")
+      println(s"Number of OSM ways: ${osm.ways.size()}")
+
+      printAllTags(osm)
+      printOSMTagInfo(osm)
+    } finally {
+      Try(osm.close())
+    }
+  }
+
+  private def printAllTags(osm: OSM): Unit = {
+    val minFrequency = 1000
+    println(s"All existing tags with the minimum frequency of $minFrequency (the format is <tag>[<frequency>]):")
+    val tagsToCnt = osm.ways.asScala
+      .flatMap {
+        case (_, way) =>
+          way.tags.asScala.map { tag =>
+            tag.key
+          }
+      }
+      .foldLeft(scala.collection.mutable.HashMap.empty[String, Int]) {
+        case (frequencyMap, tagKey) =>
+          val prevCnt = frequencyMap.getOrElse(tagKey, 0)
+          frequencyMap(tagKey) = prevCnt + 1
+          frequencyMap
+      }
+
+    val frequentTagsToCnt = tagsToCnt
+      .filter { case (_, frequency) => frequency > minFrequency }
+      .toSeq
+      .sortBy(_._2)
+      .reverse
+
+    println(frequentTagsToCnt.map { case (tag, cnt) => s"$tag [$cnt]" }.mkString("  "))
+  }
+
+  private def printOSMTagInfo(osm: OSM): Unit = {
+    val highwayType2Way: Map[String, Seq[(String, lang.Long, Way)]] = osm.ways.asScala
+      .map { way =>
+        val tags = way._2.tags
+        val highwayTag = tags.asScala.collectFirst { case tag if tag.key == "highway" => tag.value }
+        val highwayType = highwayTag.getOrElse("unknown")
+        (highwayType, way._1, way._2)
+      }
+      .toSeq
+      .groupBy { case (highwayType, _, _) => highwayType }
+
+    case class TagInfo(tagName: String, hasTag: Int, doesNotHaveATag: Int, values: Seq[String])
+    def toTagInfo(ways: Iterable[(String, lang.Long, Way)], tagToCollect: String): TagInfo = {
+      val tagVals = ways.map {
+        case (_, _, way) =>
+          if (way.hasTag(tagToCollect)) Some(way.getTag(tagToCollect))
+          else None
+      }
+      val doesNotHaveATag = tagVals.count(tv => tv.isEmpty)
+      val doesHaveATag = ways.size - doesNotHaveATag
+
+      TagInfo(tagToCollect, doesHaveATag, doesNotHaveATag, tagVals.flatten.toSeq)
+    }
+
+    val allTags = Set("lanes", "maxspeed", "capacity")
+
+    println("The number of tags present for each way, grouped by highway type.")
+    val highwayType2TagInfo = highwayType2Way.map {
+      case (highwayType, ways) =>
+        val tagInfos = allTags.map { tagToAnalyze =>
+          toTagInfo(ways, tagToAnalyze)
+        }
+        highwayType -> tagInfos
+    }
+
+    highwayType2TagInfo.foreach {
+      case (highwayType, tagInfos) =>
```

Comment:
Seems like this is a long anonymous function and it's not fully clear what it is doing. Maybe we could extract some methods with meaningful names?

Reply:
to be honest - I do not see any simple way to make code to print things better

Reply:
Now it seems much better